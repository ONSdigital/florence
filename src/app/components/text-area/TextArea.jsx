import React from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import Panel from "../panel/Panel";

const propTypes = {
    name: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    handleChange: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired,
    descriptionHint: PropTypes.string,
    error: PropTypes.string,
    maxLength: PropTypes.number,
    placeholder: PropTypes.string,
    rows: PropTypes.number,
};

const TextArea = props => {
    const charactersRemaining = props.maxLength - props.text.length;
    const textAreaBody = (
        <div className={clsx("form__input", { "form__input--error": props.error })}>
            <label htmlFor={props.name} className="form__label">
                {props.title}
            </label>
            <div>{props.descriptionHint}</div>
            <textarea
                id={props.name}
                name={props.name}
                className="input input__textarea"
                value={props.text}
                placeholder={props.placeholder}
                aria-describedby={`${props.name}-char-limit-remaining`}
                rows={props.rows}
                maxLength={props.maxLength}
                onChange={e => props.handleChange(e.target.value)}
            />
            {props.maxLength && (
                <div id={`${props.name}-char-limit-remaining`}>
                    <strong>You have {charactersRemaining} characters remaining</strong>
                </div>
            )}
        </div>
    );

    // If there is an error then render the text-area within an error panel
    if (props.error) {
        return <Panel type="error" heading={props.error} body={textAreaBody} />;
    }
    return textAreaBody;
};

TextArea.propTypes = propTypes;

export default TextArea;
