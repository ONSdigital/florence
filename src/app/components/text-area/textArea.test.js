import React from "react";
import TextArea from "./TextArea";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("TextArea", () => {
    describe("Given props that do not include the error attribute", () => {
        const textAreaProps = {
            title: "Please provide some feedback",
            name: "feedback",
            text: "",
            descriptionHint: "For example, describe any difficulties you experienced in the use of this service",
            maxLength: 200,
            placeholder: "Please provide your feedback",
            rows: 8,
            error: "",
            handleChange: () => {},
        };
        const component = shallow(<TextArea {...textAreaProps} />);
        it("Should render the textArea", () => {
            expect(component.find("#feedback").text()).toBe(textAreaProps.text);
            expect(component.find(".form__label").text()).toBe(textAreaProps.title);
            expect(component.find(".panel--error").length).toBe(0);
            expect(component.find("#feedback-char-limit-remaining").length).toBe(1);
            expect(component.find(".form__input--error").length).toBe(0);
        });
        it("matches the snapshot", () => {
            const componentForSnapshot = renderer.create(<TextArea {...textAreaProps} />);
            expect(componentForSnapshot.toJSON()).toMatchSnapshot();
        });
    });
    describe("Given props which include the error attribute", () => {
        const textAreaProps = {
            title: "Please provide some feedback",
            name: "feedback",
            text: "Foo",
            descriptionHint: "For example, describe any difficulties you experienced in the use of this service",
            maxLength: 200,
            placeholder: "Please provide your feedback",
            rows: 8,
            error: "Something went wrong",
            handleChange: () => {},
        };
        it("matches the snapshot", () => {
            const componentForSnapshot = renderer.create(<TextArea {...textAreaProps} />);
            expect(componentForSnapshot.toJSON()).toMatchSnapshot();
        });
    });
});
