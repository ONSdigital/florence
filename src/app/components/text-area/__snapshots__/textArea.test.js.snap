// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Given props that do not include the error attribute matches the snapshot 1`] = `
<div
  className="form__input"
>
  <label
    className="form__label"
    htmlFor="feedback"
  >
    Please provide some feedback
  </label>
  <div>
    For example, describe any difficulties you experienced in the use of this service
  </div>
  <textarea
    aria-describedby="feedback-char-limit-remaining"
    className="input input__textarea"
    id="feedback"
    maxLength={200}
    name="feedback"
    onChange={[Function]}
    placeholder="Please provide your feedback"
    rows={8}
    value=""
  />
  <div
    id="feedback-char-limit-remaining"
  >
    <strong>
      You have 
      200
       characters remaining
    </strong>
  </div>
</div>
`;

exports[`TextArea Given props which include the error attribute matches the snapshot 1`] = `
<div
  className="error"
>
  <div
    className=""
  >
    <div
      className="panel panel--error panel--no-title"
    >
      <span
        className="visually-hidden"
      >
        error
      </span>
      <div
        className="panel__body"
      >
        <p
          className="panel__error"
        >
          <strong
            className="font-weight--600"
          >
            Something went wrong
          </strong>
        </p>
        <div
          className="form__input form__input--error"
        >
          <label
            className="form__label"
            htmlFor="feedback"
          >
            Please provide some feedback
          </label>
          <div>
            For example, describe any difficulties you experienced in the use of this service
          </div>
          <textarea
            aria-describedby="feedback-char-limit-remaining"
            className="input input__textarea"
            id="feedback"
            maxLength={200}
            name="feedback"
            onChange={[Function]}
            placeholder="Please provide your feedback"
            rows={8}
            value="Foo"
          />
          <div
            id="feedback-char-limit-remaining"
          >
            <strong>
              You have 
              197
               characters remaining
            </strong>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
